internal/encryption/keyprovider/openbao/client.go

```diff
@@ -0,0 +1,96 @@
+package openbao
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	openbao "github.com/openbao/openbao/api"
+)
+
+type client interface {
+	WriteBytesWithContext(ctx context.Context, path string, data []byte) (*openbao.Secret, error)
+}
+
+// service implements missing utility functions from openbao/api such as routing and serialization.
+type service struct {
+	c client
+}
+
+type dataKey struct {
+	Plaintext  []byte
+	Ciphertext []byte
+}
+
+func (s service) generateDataKey(ctx context.Context, keyName string, bitSize int) (dataKey, error) {
+	path := fmt.Sprintf("/transit/datakey/plaintext/%s", keyName)
+
+	reqBody, err := json.Marshal(map[string]interface{}{
+		"bits": bitSize,
+	})
+	if err != nil {
+		return dataKey{}, fmt.Errorf("serializing datakey request to openbao: %w", err)
+	}
+
+	secret, err := s.c.WriteBytesWithContext(ctx, path, reqBody)
+	if err != nil {
+		return dataKey{}, fmt.Errorf("sending datakey request to openbao: %w", err)
+	}
+
+	key := dataKey{}
+
+	key.Ciphertext, err = retrieveCiphertext(secret)
+	if err != nil {
+		return dataKey{}, err
+	}
+
+	key.Plaintext, err = retrievePlaintext(secret)
+	if err != nil {
+		return dataKey{}, err
+	}
+
+	return key, nil
+}
+
+func (s service) decryptData(ctx context.Context, keyName string, ciphertext []byte) ([]byte, error) {
+	path := fmt.Sprintf("/transit/decrypt/%s", keyName)
+
+	reqBody, err := json.Marshal(map[string]string{
+		"ciphertext": string(ciphertext),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("serializing decryption request to openbao: %w", err)
+	}
+
+	secret, err := s.c.WriteBytesWithContext(ctx, path, reqBody)
+	if err != nil {
+		return nil, fmt.Errorf("sending decryption request to openbao: %w", err)
+	}
+
+	return retrievePlaintext(secret)
```

Comment:
Can we please handle the error before directly returning the plaintext?

Reply:
That already happens in the implementation of retrievePlaintext.  As this is a private function, I don't believe that is necessary here.

Reply:
As this is already handled by retrievePlaintext and is a private function, I don't see the need for that here.